compiler/sigmatch.nim

```diff
@@ -2276,14 +2287,18 @@ proc prepareOperand(c: PContext; formal: PType; a: PNode): PNode =
     # a.typ == nil is valid
     result = a
   elif a.typ.isNil:
-    # XXX This is unsound! 'formal' can differ from overloaded routine to
-    # overloaded routine!
-    let flags = {efDetermineType, efAllowStmt}
-                #if formal.kind == tyIter: {efDetermineType, efWantIterator}
-                #else: {efDetermineType, efAllowStmt}
-                #elif formal.kind == tyTyped: {efDetermineType, efWantStmt}
-                #else: {efDetermineType}
-    result = c.semOperand(c, a, flags)
+    if formal.kind == tyIterable:
```

Comment:
Why would `a.typ` be nil here for this case?

Reply:
because it's not type-checked yet, the type-checking happens inside `semOperand`, nothing unusual here.

this would be the same with `myiter()` or `123` below.

```nim
when true:
  iterator myiter(): int =
    yield 10
  template baz1(a: iterable[int]) = discard
  baz1(myiter())

  template baz2(a: int) = discard
  baz2(123)
```